components/ws-daemon/pkg/cpulimit/cpulimit.go

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package cpulimit
+
+// TODO(cw):
+//    - introduce metrics that report
+// 			- total throttled period count (per QoS)
+//			- total bandwidth granted (per QoS)
+//			- something that corresponds to the user experience
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type Workspace struct {
+	ID string
+
+	NrThrottled uint64
+	Usage       CPUTime
+	QoS         int
+}
+
+type WorkspaceHistory struct {
+	ID string
+
+	LastUpdate  *Workspace
+	UsageT0     CPUTime
+	ThrottleLag uint64
+	Limit       Bandwidth
+}
+
+func (h *WorkspaceHistory) Usage() CPUTime {
+	if h == nil || h.LastUpdate == nil {
+		panic("usage called before update")
```

Comment:
should we panic here? can this be an error instead? panic would mean a hard crash, which might not be great for ws-daemon